ambry-server/src/main/java/com.github.ambry.server/StatsManager.java

```diff
@@ -143,22 +144,59 @@ void collectAndAggregate(StatsSnapshot aggregatedSnapshot, PartitionId partition
   }
 
   /**
-   * Performs recursive aggregation of two {@link StatsSnapshot} and stores the result in the first one.
-   * @param baseSnapshot one of the addends and where the result will be
-   * @param newSnapshot the other addend to be added into the first {@link StatsSnapshot}
+   * Fetch the {@link StatsSnapshot} for the given {@link PartitionId}.
+   * @param partitionId the {@link PartitionId} to try to fetch the {@link StatsSnapshot} from
+   * @param unreachableStores a list of partitionIds to keep track of the unreachable stores (partitions)
+   * @return
    */
-  private void aggregate(StatsSnapshot baseSnapshot, StatsSnapshot newSnapshot) {
-    baseSnapshot.setValue(baseSnapshot.getValue() + newSnapshot.getValue());
-    if (baseSnapshot.getSubMap() == null) {
-      baseSnapshot.setSubMap(newSnapshot.getSubMap());
-    } else if (newSnapshot.getSubMap() != null) {
-      for (Map.Entry<String, StatsSnapshot> entry : newSnapshot.getSubMap().entrySet()) {
-        if (!baseSnapshot.getSubMap().containsKey(entry.getKey())) {
-          baseSnapshot.getSubMap().put(entry.getKey(), new StatsSnapshot(0L, null));
+  StatsSnapshot fetchSnapshot(PartitionId partitionId, List<String> unreachableStores) {
+    StatsSnapshot statsSnapshot = null;
+    Store store = storageManager.getStore(partitionId);
+    if (store == null) {
+      unreachableStores.add(partitionId.toString());
+    } else {
+      try {
+        statsSnapshot = store.getStoreStats().getStatsSnapshot();
+      } catch (StoreException e) {
+        unreachableStores.add(partitionId.toString());
+      }
+    }
+    return statsSnapshot;
+  }
+
+  /**
+   * Get the combined {@link StatsSnapshot} of all partitions in this node.
+   * @return a combined {@link StatsSnapshot} of this node
```

Comment:
some example on how a combined would look like. Easy to get confused that we might reduce it to just one map w/o per partition info